---
title: "Single cell RNA-seq of Plasmodium sporozoite-infected monocytes - demultiplexing"
output:
  pdf_document: default
  html_document:
    df_print: paged
  toc: TRUE
---

<!-- ```{r, echo = F} -->
<!-- # knitr::opts_chunk$set(warning = F, message = F) -->
<!-- # knitr::opts_chunk$set(fig.pos = "!H", out.extra = "") -->
<!-- #knitr::opts_chunk$set(tidy.opts=list(width.cutoff=80),tidy=TRUE) -->

<!-- ``` -->

<!-- ```{r load_libraries, echo = F, message = F, warning = F} -->
<!-- library(Seurat) -->
<!-- library(patchwork) -->
<!-- library(readxl) -->
<!-- library(tidyverse) -->
<!-- library(ggplot2) -->
<!-- library(ggpubr) -->
<!-- library(reshape2) -->
<!-- library(broom) -->
<!-- library(stringr) -->
<!-- # remotes::install_github('chris-mcginnis-ucsf/DoubletFinder') -->
<!-- # library(DoubletFinder) -->
<!-- ``` -->

<!-- ```{r} -->
<!-- load("malaria_monocyte_QC.RData") -->
<!-- ``` -->

# Step 2: Demultiplexing 

<!-- a) match donor ID with samples (cells)  -->
<!-- ```{r demultiplexing_vireo} -->

<!-- # install / upgrade vireoSNP using pip install --upgrade --no-deps vireoSNP -->
<!-- # vireo to check installation errors -->
<!-- # download all cellSNP file from library -> save vcf file as .gz, leave others un-gzipped -->
<!-- # error: index out of range -> solved by downloading the files again and redoing vireo -->
<!-- # command line: vireo -c cellSNP_mat/ -N 5 -o vireo_result/ -->

<!-- # add vireo output to meta data -->
<!-- mono.demul <- mono -->
<!-- snp <- read.delim(paste("../counts/",lib_name,"/vireo_result/donor_ids.tsv", sep = "")) -->
<!-- meta <- mono.demul@meta.data %>%  -->
<!--   tibble::rownames_to_column("cell") %>%  -->
<!--   left_join(snp) -->

<!-- table(meta$donor_id) -->
<!-- # remove doublets and unassigned -->
<!-- mono.demul@meta.data <- meta -->

<!-- Idents(mono.demul) <- "donor_id" -->
<!-- mono.singlet.donor <- mono.demul@meta.data %>%  -->
<!--   as.data.frame() %>%  -->
<!--   tibble::column_to_rownames("cell") -->

<!-- mono.demul@meta.data <- mono.singlet.donor -->

<!-- mono.singlet.donor <- subset(x = mono.demul, idents = c("doublet", "unassigned"), invert = T) -->

<!-- # identical(meta$cell, rownames(mono@meta.data)) -->
<!-- # [1] TRUE -->
<!-- ``` -->




PCA on singlets
```{r mono.singlet}

# ##

# Select the top 1000 most variable features
mono <- FindVariableFeatures(mono)

# Scaling RNA data, we only scale the variable features here for efficiency
mono <- ScaleData(mono, features = VariableFeatures(mono))

# Run PCA
mono <- RunPCA(mono, features = VariableFeatures(mono))

# We select the top 10 PCs for clustering and tSNE based on PCElbowPlot
mono <- FindNeighbors(mono, reduction = "pca", dims = 1:10)
mono <- FindClusters(mono, resolution = 0.6, verbose = FALSE)
mono <- RunTSNE(mono, reduction = "pca", dims = 1:10)

# Projecting singlet identities on TSNE visualization
#DimPlot(mono, group.by = "HTO_classification")

#DimPlot(mono, group.by = "HTO_classification")
DimPlot(mono, group.by = "MULTI_classification")
```

singlets, doublets, negatives
```{r dem_QC}
#mono <- subset(mono, subset = nFeature_RNA > 200 & nFeature_RNA < 2500 & percent.mt < 5)

# Visualize QC metrics as a violin plot
#VlnPlot(mono.singlet, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

mono@meta.data %>% 
  ggplot(., aes(x = nFeature_RNA)) +
  geom_histogram(binwidth = 30)

mono@meta.data %>% 
  ggplot(., aes(x = nCount_RNA)) +
  geom_histogram(binwidth = 70)

mono@meta.data %>% 
  ggplot(., aes(x = percent.mt)) +
  geom_histogram(binwidth = 0.11)

mono@meta.data %>% 
  ggplot(., aes(x = percent.plasmo)) +
  geom_histogram(binwidth = 0.11)

# FeatureScatter is typically used to visualize feature-feature relationships, but can be used
# for anything calculated by the object, i.e. columns in object metadata, PC scores etc.

plot1 <- FeatureScatter(mono, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(mono, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1 + plot2
```

```{r}
mono.demul <- mono
```


```{r}
#save.image("malaria_monocyte_demultiplex.RData")

```