---
title: "QC_Demultiplex_Naive_custom_cellSNP"
output: html_notebook
---

```{r opts1, echo = F}
knitr::opts_chunk$set(warning = F, message = F)
knitr::opts_chunk$set(fig.width=12, fig.height=8) 

```

# Step 1: Load filtered features, barcode, matrices and visualise quality


```{r datadiscovery, echo = F}

samplesheet <- read_excel("../Data/a_2024_HUSTLEseq_scRNAseq_nucleotide_sequencing_round2.xlsx")

colnames(samplesheet) <- sapply(c(1:ncol(samplesheet)), function(x) paste0(samplesheet[1,x], 
                                                                           colnames(samplesheet)[x], 
                                                                           collapse = ""))

samplesheet <- samplesheet %>% 
  filter(!row_number() %in% c(1)) %>% 
  dplyr::rename(Library_name = `NameLibrary Name`) %>% 
  dplyr::rename(Antibody = `Multiplex Source Tag...31`)

lib_name = "D13_D33_D45_D47_CD4_Naive"
d <- samplesheet %>% filter(Library_name == lib_name)


# Antibody column needs to be changed from Total_Seq to TotalSeq
samplesheet <- samplesheet %>%
  mutate(Antibody = gsub("Total_Seq", "TotalSeq", Antibody))
```

The hastag antibodies used were:
```{r hto_used}

# lib name
hto_used = samplesheet %>%
    filter(Library_name == lib_name) %>%
    pull(Antibody) %>%
    unique

hto_used
```


```{r}

# for Naive, to include more cells from raw features 
seurat_filtered <- readRDS("robjects/D13_D33_D45_D47_CD4_Naive_filtered_17k.rds")
# a. Filter by min.features (per cell):

seurat_filtered <- subset(seurat_filtered, subset = nFeature_RNA >= 200)

#########into cellSNP
#raw_features <- as.data.frame(colnames(seurat_filtered))
#write_tsv(raw_features, "hustleseq_naive_cellbarcodes.tsv")

# b. Filter by min.cells (per gene):

# Count how many cells each gene is expressed in
gene_cell_counts <- rowSums(seurat_filtered[["RNA"]]@layers$counts > 0)

# Keep genes expressed in at least 3 cells
genes_to_keep <- names(gene_cell_counts[gene_cell_counts >= 3])

# Subset RNA assay
hustle <- subset(seurat_filtered, features = genes_to_keep)
```


```{r}
hustle[["percent.mt"]] <- PercentageFeatureSet(hustle, pattern = "^MT-")
ribo.genes <- grep("^RP[SL][[:digit:]]|^RPLP[[:digit:]]|^RPSA",rownames(hustle),value = TRUE)
ribo.genes <- grep("[BK-]|AP|[[:digit:]](.+)L",ribo.genes,value = TRUE, invert = T)
hustle[["percent.ribo"]] <- PercentageFeatureSet(object = hustle, features = ribo.genes)
head(hustle@meta.data, 5)

```

Demultiplex: Assign donor to cells - singlets, doublets, negatives using vireo

a) match donor ID with samples (cells) 
```{r demultiplexing_vireo}

# install / upgrade vireoSNP using pip install --upgrade --no-deps vireoSNP
# vireo to check installation errors
# download all cellSNP file from library -> save vcf file as .gz, leave others un-gzipped
# error: index out of range -> solved by downloading the files again and redoing vireo
# command line: vireo -c cellSNP_mat/ -N 3 -o vireo_result/ for 3 donors

# add vireo output to meta data
hustle.demul <- hustle
snp <- read.delim(paste("../Data/",lib_name,"/vireo_result_rerun_PM/donor_ids.tsv", sep = ""))
meta <- hustle.demul@meta.data %>% 
  tibble::rownames_to_column("cell") %>% 
  left_join(snp)

table(meta$donor_id)
# remove doublets and unassigned
hustle.demul@meta.data <- meta

hustle.demul@meta.data <- hustle.demul@meta.data %>% 
  as.data.frame() %>% 
  tibble::column_to_rownames("cell")

## didn't need the next line for CellBender output:
Idents(hustle.demul) <- "donor_id"
VlnPlot(hustle.demul, features = c("nFeature_RNA", "nCount_RNA"), pt.size = 0.1, log = F)
ggplot(hustle.demul@meta.data, aes(x = nCount_RNA)) + geom_histogram(binwidth = 50)
ggplot(hustle.demul@meta.data, aes(x = nCount_RNA)) + 
  geom_histogram(binwidth = 50) +
  facet_wrap( ~ donor_id) +
  theme_bw()

hustle.demul <- subset(x = hustle.demul, idents = c("doublet", "unassigned"), invert = T)

hustle <- hustle.demul

hustle <- subset(hustle, subset = !is.na(donor_id))

VlnPlot(hustle, features = c("nFeature_RNA", "nCount_RNA"), pt.size = 0.1, log = F)
ggplot(hustle@meta.data, aes(x = nCount_RNA)) + geom_histogram(colour = "white", bins = 50) + theme_bw()
ggplot(hustle@meta.data, aes(x = nCount_RNA)) + 
  geom_histogram(colour = "white", bins = 50) +
  facet_wrap( ~ donor_id) +
  theme_bw()

ggplot(hustle@meta.data, aes(x = nFeature_RNA)) + 
  geom_histogram(colour = "white", bins = 50) +
  facet_wrap( ~ donor_id) +
  theme_bw()
```

b) Demultiplexing with hastag oligos -> identify cells with tagged barcodes (# of barcodes (Ab) = # of donors)
```{r Demultiplexing_HTO}
#Cell Hashing uses oligo-tagged antibodies against ubuquitously expressed surface proteins to place a “sample barcode” on each single cell, enabling different samples to be multiplexed together and run in a single experiment.

# Select cell barcodes detected by both RNA and HTO In the example datasets we have already
# filtered the cells for you, but perform this step for clarity.

hustle.singlet.donor <- hustle
joint.bcs <- rownames(hustle.singlet.donor@meta.data)

# Subset RNA and HTO counts by joint cell barcodes
hustle.hto <- as.matrix(hustle[["HTO"]]@counts[,joint.bcs])

## # Add HTO data as a new assay independent from RNA
Idents(hustle.singlet.donor) <- rownames(hustle.singlet.donor@meta.data)
hustle.singlet.donor[["HTO"]] <- CreateAssayObject(counts = hustle.hto)

# Normalize HTO data, here we use centered log-ratio (CLR) transformation
hustle.singlet.donor <- NormalizeData(hustle.singlet.donor, assay = "HTO", normalization.method = "CLR")


# If you have a very large dataset we suggest using k_function = 'clara'. This is a k-medoid
# clustering function for large applications You can also play with additional parameters (see
# documentation for HTODemux()) to adjust the threshold for classification Here we are using
# the default settings
hustle.singlet.donor <- HTODemux(hustle.singlet.donor, assay = "HTO", positive.quantile = 0.99) #, positive.quantile = 0.99
RidgePlot(hustle.singlet.donor, assay = "HTO", features = rownames(hustle.singlet.donor[["HTO"]]), ncol = 3)
```


```{r plots_dem}

# Global classification results
table(hustle.singlet.donor$HTO_classification.global)

Idents(hustle.singlet.donor) <- "HTO_classification.global"
hustle.h <- hustle.singlet.donor

if(any(hustle.h$HTO_classification.global == "Negative") == T)
   hustle.h <- subset(hustle.h, idents = c("Negative"), invert = TRUE)


#Extract singlets
hashtags <- gsub("Total_Seq", "TotalSeq", hto_used)
hashtags <- gsub("_", "-", hto_used)

hustle.h@meta.data <- hustle.h@meta.data %>%
  mutate(Cell_state = case_when(
    as.character(HTO_classification.global) %in% hashtags ~ "Singlet",
    .default = HTO_classification.global
  ))

Idents(hustle.h) <- "Cell_state"

VlnPlot(hustle.h, features = c("nFeature_RNA", "nCount_RNA"), pt.size = 0.1, log = F)
ggplot(hustle.singlet.donor@meta.data, aes(x = nCount_RNA)) + geom_histogram(binwidth = 50)
ggplot(hustle.singlet.donor@meta.data, aes(x = nCount_RNA)) + 
  geom_histogram(binwidth = 50) +
  facet_wrap( ~ HTO_classification.global) +
  theme_bw()

ggplot(hustle.singlet.donor@meta.data, aes(x = nFeature_RNA)) + geom_histogram(binwidth = 50)
ggplot(hustle.singlet.donor@meta.data, aes(x = nFeature_RNA)) + 
  geom_histogram(binwidth = 50) +
  facet_wrap( ~ HTO_classification.global) +
  theme_bw()

hustle.h <- subset(hustle.h, subset = Cell_state == "Singlet")

```

Features and samples after assigning cells to donors and retaining only singlets
```{r}
Idents(hustle.singlet.donor) = "HTO_classification.global"

# For HTODemux:
hustle.singlet.donor = hustle.h
Idents(hustle.singlet.donor) = "HTO_classification.global"
hustle.h <- subset(hustle.singlet.donor, idents = "Singlet")
hustle.h

# Calculate a tSNE embedding of the HTO data
DefaultAssay(hustle.h) <- "HTO"
hustle.h <- ScaleData(hustle.h, features = rownames(hustle.h),
    verbose = FALSE)
hustle.h <- RunPCA(hustle.h, features = rownames(hustle.h), approx = FALSE)
hustle.h <- RunTSNE(hustle.h, check_duplicates = F)

hustle <- hustle.h

```

QC for number of RNA molecules (nCount_RNA), number of features (nFeature_RNA), mitochondrial RNA percent denoting dead cells (percent.mt) and ribosomal genes percentage (NA cells because of unmatched CellBender cell filtration)

```{r Visviolin}
# Visualize QC metrics as a violin plot
DefaultAssay(hustle) <- "RNA"
Idents(hustle) <- "orig.ident"
VlnPlot(hustle, features=c("nCount_RNA", "nFeature_RNA", "percent.mt"), ncol = 3) 
VlnPlot(hustle, features=c("percent.ribo")) 

ggplot(hustle@meta.data, aes(x = nFeature_RNA)) + geom_histogram(colour = "white", bins = 50)

ggplot(hustle@meta.data, aes(x = nFeature_RNA)) + 
  geom_histogram(binwidth = 50) +
  facet_wrap( ~ donor_id)

ggplot(hustle@meta.data, aes(x = nCount_RNA)) + geom_histogram(colour = "white",bins = 50)
ggplot(hustle@meta.data, aes(x = nCount_RNA)) + 
  geom_histogram(binwidth = 50) +
  facet_wrap( ~ donor_id)

ggplot(hustle@meta.data, aes(x = percent.mt)) + geom_histogram(colour = "white",bins = 50)
ggplot(hustle@meta.data, aes(x = percent.mt)) + 
  geom_histogram(binwidth = 1) +
  facet_wrap( ~ donor_id)

ggplot(hustle@meta.data, aes(x = percent.ribo)) + geom_histogram(colour = "white",bins = 50)
ggplot(hustle@meta.data, aes(x = percent.ribo)) + 
  geom_histogram(binwidth = 1) +
  facet_wrap( ~ donor_id)

```

```{r}
# make UMI counts vs barcodes plot

# Extract UMI counts - raw
umi_counts_raw <- hustle.h@meta.data$nCount_RNA
umi_counts_raw <- sort(umi_counts_raw, decreasing = TRUE)  # Sort in descending order

# Prepare data for plotting
umi_data_raw <- data.frame(
  Rank = 1:length(umi_counts_raw),
  UMI = umi_counts_raw
)

threshold <- 700  # Example threshold
p1 <- ggplot(umi_data_raw, aes(x = Rank, y = UMI)) +
  geom_point(size = 0.5) +
  geom_hline(yintercept = threshold, linetype = "dashed", color = "red") +
  annotate(
    "text", 
    x = max(umi_data_raw$Rank) * 0.8,  # Adjust x position (80% of the x-axis range)
    y = threshold, 
    label = paste("Threshold =", threshold),
    color = "red",
    hjust = 0,  # Horizontal alignment of the text
    vjust = -0.5  # Vertical alignment of the text (above the line)
  ) +
  # scale_y_continuous( limits=c(0,max(umi_data_raw$UMI)), 
  #   breaks = c(1, 10, 100, 1000, 10000),
  #                    labels = c("1", "10", "100", "1000", "10000")) +
  scale_y_log10(
    breaks = c(300, 1000, 3000, 10000),  # Specify desired ticks
    labels = c("300", "1000", "3000", "10000")
  ) +
  scale_x_log10() +
  labs(
    x = "Barcodes (ranked)",
    y = "UMI Counts",
    title = "UMI Counts vs. Barcodes"
  ) +
  theme_minimal()
p1
#ggsave("UMI counts vs barcodes raw and filtered features.pdf")
```

```{r}
deadcell_nF_lowercutoff <- 150
deadcell_nF_uppercutoff <- 1200
deadcell_nC_lowercutoff <- 500
deadcell_mt_cutoff <- 12

hustle.meta <- hustle@meta.data %>% 
  mutate(is.dead = case_when(
      nFeature_RNA >= deadcell_nF_lowercutoff & 
        nFeature_RNA <= deadcell_nF_uppercutoff &
        nCount_RNA >= deadcell_nC_lowercutoff &
        percent.mt <= deadcell_mt_cutoff ~ "FALSE",
      TRUE ~ "TRUE"))

table(hustle.meta$is.dead)

ggplot(hustle.meta, aes(x = nFeature_RNA, fill = factor(is.dead))) +
  geom_histogram(bins = 50, position = "identity", alpha = 0.5)

ggplot(hustle.meta, aes(x = nCount_RNA, fill = factor(is.dead))) +
  geom_histogram(bins = 50, position = "identity", alpha = 0.5)

ggplot(hustle.meta, aes(x = percent.mt, fill = factor(is.dead))) +
  geom_histogram(bins = 50, position = "identity", alpha = 0.5)

ggplot(hustle.meta, aes(x = percent.ribo, fill = factor(is.dead))) +
  geom_histogram(bins = 50, position = "identity", alpha = 0.5)
```

```{r}
ggplot(hustle.meta, aes(x = nFeature_RNA, y = nCount_RNA)) +
  geom_point() +
  theme_bw()
```

```{r filtering}
hustle1 <- subset(hustle, subset =  nFeature_RNA >= deadcell_nF_lowercutoff & 
                    nFeature_RNA <= deadcell_nF_uppercutoff & 
                    percent.mt <= deadcell_mt_cutoff &
                    nCount_RNA >= deadcell_nC_lowercutoff 
                    ) #nCount_RNA >= 1000 &
hustle1
VlnPlot(hustle1, features=c("nCount_RNA", "nFeature_RNA", "percent.mt", "percent.ribo"), group.by = "orig.ident", ncol = 2) 

hustle = hustle1

```


```{r}
# make UMI counts vs barcodes plot

# Extract UMI counts - raw
umi_counts_raw <- hustle@meta.data$nCount_RNA
umi_counts_raw <- sort(umi_counts_raw, decreasing = TRUE)  # Sort in descending order

# Prepare data for plotting
umi_data_raw <- data.frame(
  Rank = 1:length(umi_counts_raw),
  UMI = umi_counts_raw
)

threshold <- 700  # Example threshold
p1 <- ggplot(umi_data_raw, aes(x = Rank, y = UMI)) +
  geom_point(size = 0.5) +
  geom_hline(yintercept = threshold, linetype = "dashed", color = "red") +
  annotate(
    "text", 
    x = max(umi_data_raw$Rank) * 0.8,  # Adjust x position (80% of the x-axis range)
    y = threshold, 
    label = paste("Threshold =", threshold),
    color = "red",
    hjust = 0,  # Horizontal alignment of the text
    vjust = -0.5  # Vertical alignment of the text (above the line)
  ) +
  # scale_y_continuous( limits=c(0,max(umi_data_raw$UMI)), 
  #   breaks = c(1, 10, 100, 1000, 10000),
  #                    labels = c("1", "10", "100", "1000", "10000")) +
  scale_y_log10(
    breaks = c(300, 1000, 3000, 10000),  # Specify desired ticks
    labels = c("300", "1000", "3000", "10000")
  ) +
  scale_x_log10() +
  labs(
    x = "Barcodes (ranked)",
    y = "UMI Counts",
    title = "UMI Counts vs. Barcodes"
  ) +
  theme_minimal()
p1
#ggsave("UMI counts vs barcodes raw and filtered features.pdf")
```


Assigning hashtags and peptide pools to cells
```{r}
# assigning cell to donor and pool

hustle@meta.data <- hustle@meta.data %>% 
  mutate(Donor_expt = case_when(
    HTO_classification == "TotalSeq-C-1" ~ "D13",
    HTO_classification == "TotalSeq-C-2" ~ "D33",
    HTO_classification == "TotalSeq-C-3" ~ "D45",
    HTO_classification == "TotalSeq-C-4" ~ "D47",
    .default = ""
  )) %>% 
  mutate(pep.pool = case_when(
    HTO_classification == "TotalSeq-C-5" ~ "P2",
    HTO_classification == "TotalSeq-C-6" ~ "P3",
    HTO_classification == "TotalSeq-C-7" ~ "P4",
    HTO_classification == "TotalSeq-C-8" ~ "P5",
    HTO_classification == "TotalSeq-C-9" ~ "P6",
    HTO_classification == "TotalSeq-C-10" ~ "P7",
    HTO_classification == "TotalSeq-C-11" ~ "P8",
    HTO_classification == "TotalSeq-C-12" ~ "P9",
    HTO_classification == "TotalSeq-C-2" ~ "P1",
    HTO_classification == "TotalSeq-C-3" ~ "P1",
    HTO_classification == "TotalSeq-C-4" ~ "P1",
    HTO_classification == "TotalSeq-C-1" ~ "P1",
    .default = ""
  ))

# get donor id from SNP for assigning D33, D45, D47

df <- hustle@meta.data
df <- df[df$Donor_expt != "", ]
df <- df[!df$donor_id %in% c("doublet", "unassigned"), ]

donorid_snp_hto <- dplyr::count(df, Donor_expt, donor_id)

#Put them back in hustle meta data

hustle@meta.data <- hustle@meta.data %>% 
  mutate(Donor_expt = case_when(
    HTO_classification == "TotalSeq-C-1" ~ "D13",
    HTO_classification == "TotalSeq-C-2" ~ "D33",
    HTO_classification == "TotalSeq-C-3" ~ "D45",
    HTO_classification == "TotalSeq-C-4" ~ "D47",
    donor_id == "donor0" ~ "D33",
    donor_id == "donor3" ~ "D47",
    donor_id == "donor2" ~ "D13",
    donor_id == "donor1" ~ "D45",
    .default = ""
  )) %>% 
  filter(!is.na(donor_id)) %>% 
  mutate(donor_pep.pool = paste(Donor_expt, pep.pool, sep = "_"))


table(hustle$Donor_expt)
table(hustle$pep.pool)
table(hustle$Donor_expt, hustle$pep.pool)
table(hustle$donor_pep.pool)
table(hustle$donor_id, hustle$Donor_expt)
table(hustle$HTO_classification, hustle$Donor_expt)
table(hustle$HTO_classification, hustle$donor_id)

```